### PR TITLE
msg/async/dpdk: Add padding to message which smaller than ETHER_MIN_LEN

### PR DESCRIPTION
**Pull request by rosinL**
Message which smaller than minimun frame len maybe dropped by driver. To
ensure successful transmission, add padding to the message before sending.

Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>
Signed-off-by: luo rixin <luorixin@huawei.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
> Message which smaller than minimun frame len maybe dropped by driver

by pmd driver or by DPDK? could you point me to the code or the document, or anything that support this?

**Comment by fengchunsong** [comment]
> > Message which smaller than minimun frame len maybe dropped by driver
> 
> by pmd driver or by DPDK? could you point me to the code or the document, or anything that support this?

The dpdk driver has two types of ultra-short packet processing, one is to send successfully after adding padding, and the other is to directly discard, such as
mlx5 / mlx5_rxtx.c:
3172         /*
3173          * Some Tx offloads may cause an error if
3174          * packet is not long enough, check against
3175          * assumed minimal length.
3176          */
3177         if (rte_pktmbuf_pkt_len(loc->mbuf) <= MLX5_ESEG_MIN_INLINE_SIZE)//MLX5_ESEG_MIN_INLINE_SIZE=18Byte
3178                 return MLX5_TXCMP_CODE_ERROR;

cxgbe / sge.c:
1123         /*
1124          * The chip min packet length is 10 octets but play safe and reject
1125          * anything shorter than an Ethernet header.
1126          */
1127         if (unlikely(m->pkt_len < RTE_ETHER_HDR_LEN)) {//RTE_ETHER_HDR_LEN=14Byte
1128 out_free:
1129                 rte_pktmbuf_free(m);
1130                 return 0;
1131         }

i40e / i40e_rxtx.c: 
1495                 /* check the size of packet */
1496                 if (m->pkt_len < I40E_TX_MIN_PKT_LEN) { //I40E_TX_MIN_PKT_LEN=17Byte
1497                         rte_errno = EINVAL;
1498                         return i;
1499                 }

File: src/msg/async/dpdk/DPDK.cc [hunk 1]
```diff
@@ -1096,6 +1096,16 @@ DPDKQueuePair::tx_buf* DPDKQueuePair::tx_buf::from_packet_zc(
   head->pkt_len = p.len();
   head->nb_segs = total_nsegs;
 
+  // Message amaller than minimum frame len maybe dropped by driver, so add padding to
```
[the remark this answers is not in the document]
Comment (fengchunsong) [reply src/msg/async/dpdk/DPDK.cc]: The length of the ACK message is zero, plus the length of the tcp, ip, and ether headers is 54. It is indeed greater than 18 bytes. There is no problem on these network cards. The sending failure problem occurred on the hns network card. The latest hns driver also added padding processing (https://github.com/DPDK/dpdk/commit/de620754a10974f810b878f637b1ed9f77ebd7d8#diff-c4781c8992d4d8f4b7d7f4e2b3715141), so there is no problem now.

**Comment by fengchunsong** [comment]
> > Message which smaller than minimun frame len maybe dropped by driver
> 
> by pmd driver or by DPDK? could you point me to the code or the document, or anything that support this?



> i think these driver are not comparing the pkt size with the minimum size defined by IEEE 802.3, where the pkt size should be greater or equal to 64.
> 
> w.r.t. mlx5_rxtx.c , https://github.com/DPDK/dpdk/blob/master/drivers/net/mlx5/mlx5_rxtx.c#L2446, how come the pkt size is smaller than MLX5_ESEG_MIN_INLINE_SIZE ?
> 
> w.r.t. sge.c, as we are sending an ethernet frame which has 2 mac addresses and 1 ether type as as its header. how come the packet could be smaller than that?
> 
> w.r.t. i40e_rxtx.c, `I40E_TX_MIN_PKT_LEN` is 17, and `2 * sizeof(mac_addr) + sizeof(ether_type)` is `12 + 2 = 14`. and let's take the crc footer into account, which is 4 bytes. so the total size would be 18. so, again, i don't think a ethernet pkt could be smaller than 18. if we are sending a packet sizing smaller than 18. that'd be a bug.
> 
> so none of these drivers drops an ether frame sizes less than 64 but greater or equal to 18.
> 
> but i agree we should pad the payload to be compliant to IEEE 802.3. if the hardware does not do this for us. but my question is, again,
> 
> > could you point me to the code or the document, or anything that support this?

The length of the ACK message is zero, plus the length of the tcp, ip, and ether headers is 54. It is indeed greater than 18 bytes. There is no problem on these network cards. The sending failure problem occurred on the hns network card. The latest hns driver also added padding processing (DPDK/dpdk@de62075#diff-c4781c8992d4d8f4b7d7f4e2b3715141), so there is no problem now.

**Comment by tchaikov** [comment]
@fengchunsong can we close this PR then? if the latest SPDK includes the DPDK containing the change of https://github.com/DPDK/dpdk/commit/de62075, please feel free to create a new PR or repurpose this PR to pick up the SPDK change .

**Comment by fengchunsong** [comment]
> @fengchunsong can we close this PR then? if the latest SPDK includes the DPDK containing the change of [DPDK/dpdk@de62075](https://github.com/DPDK/dpdk/commit/de62075), please feel free to create a new PR or repurpose this PR to pick up the SPDK change .

ok

**Pull request closed without merging**